examples/basic_timelock_ibe/backend/src/lib.rs

```diff
@@ -0,0 +1,384 @@
+        return key;
+    }
+
+    let request = VetKDPublicKeyRequest {
+        canister_id: None,
+        derivation_path: vec![DOMAIN_SEPARATOR.as_bytes().to_vec()],
+        key_id: bls12_381_test_key_1(),
+    };
+
+    let (result,) = ic_cdk::api::call::call::<_, (VetKDPublicKeyReply,)>(
+        vetkd_system_api_canister_id(),
+        "vetkd_public_key",
+        (request,),
+    )
+    .await
+    .expect("call to vetkd_public_key failed");
+
+    VetKeyPublicKey::from(result.public_key)
+}
+
+#[query(guard = is_authenticated)]
+fn get_lots() -> (OpenLotsResponse, ClosedLotsResponse) {
+    let mut open_lots = OpenLotsResponse::default();
+    let mut closed_lots = ClosedLotsResponse::default();
+
+    LOTS.with_borrow(|lots| {
+        for (lot_id, lot) in lots.iter() {
+            match lot.status {
+                LotStatus::Open => {
+                    open_lots.lots.push(lot);
+                    let bidders: Vec<Principal> = BIDS_ON_LOTS.with_borrow(|bids| {
+                        bids.range((lot_id, Principal::management_canister())..)
+                            .take_while(|((this_lot_id, _), _)| *this_lot_id == lot_id)
+                            .map(|((_, bidder), _)| bidder)
+                            .collect()
+                    });
+                    open_lots.bidders.push(bidders);
+                }
+                _ => {
+                    closed_lots.lots.push(lot);
+
+                    let bids: Vec<(Principal, u128)> = BIDS_ON_LOTS.with_borrow(|bids| {
+                        bids.range((lot_id, Principal::management_canister())..)
+                            .take_while(|((this_lot_id, _), _)| *this_lot_id == lot_id)
+                            .map(|((_, bidder), bid)| match bid {
+                                Bid::Encrypted(_) => {
+                                    panic!("bug: encrypted bid in a closed lot")
+                                }
+                                Bid::Decrypted(decrypted_bid) => (bidder, decrypted_bid.amount),
+                            })
+                            .collect()
+                    });
+                    closed_lots.bids.push(bids);
+                }
+            }
+        }
+    });
+    (open_lots, closed_lots)
+}
+
+#[update(guard = is_authenticated)]
+fn place_bid(lot_id: u128, encrypted_amount: Vec<u8>) -> Result<(), String> {
+    let bidder = ic_cdk::caller();
+    let now = ic_cdk::api::time();
+
+    LOTS.with_borrow(|lots| match lots.get(&lot_id) {
+        Some(LotInformation {
+            status: LotStatus::Open,
+            creator,
+            end_time,
+            ..
+        }) if creator != bidder && now < end_time => Ok(()),
+        Some(LotInformation { creator, .. }) if creator == bidder => {
+            Err("lot creator cannot bid".to_string())
+        }
+        Some(_) => Err("lot is closed".to_string()),
+        None => Err("lot not found".to_string()),
+    })?;
```

Comment:
IIUC the same bidder can make multiple bids. Which is maybe OK/expected; their highest bid is what counts in the end. I'm just thinking that `close_one_lot_if_any_is_open` may run into problems if there are millions of bids on one lot, and decrypting the lot takes more than 5 seconds. [I don't know how the runtime reacts to this situation, where a timer goes off and the function is still running when the next timer round occurs]

Reply:
> We overwrite user's bid if passed multiple ones and only the last one counts. So you would need many principals for that.

OK I had missed this part. I thought if a single principal spammed orders we would not dedup.